File: Episodio04 - No usar null/Ruby/Ejemplo-AddressNoPuedeSerNil/address.rb
class Address
  INVALID_STREET = 'Street can not be empty'
  INVALID_CITY = 'City can not be empty'

  @street
  @city
  @zip_code

  def self.at(street, city, zip_code)
    raise Exception.new INVALID_STREET if street.empty?
    raise Exception.new INVALID_CITY if city.empty?

    new street,city,zip_code
  end

  def initialize(street, city, zip_code)
    @street = street
    @city = city
    @zip_code = zip_code
  end

  def city
    @city
  end

  def zip_code
    @zip_code
  end

end

File: Episodio04 - No usar null/Java/Episodio04-Ejemplo-NullObject-ComoObjeto/src/Address.java

public class Address {

	public static final String INVALID_STREET = "Street can not be empty";
	public static final String INVALID_CITY = "City can not be empty";
	public static final Address NOT_PROVIDED = new Address("","","");
	
	private String street;
	private String city;
	private String zipCode;

	public static Address at(String street, String city, String zipCode) {
		if(street.isEmpty()) throw new RuntimeException(INVALID_STREET);
		if(city.isEmpty()) throw new RuntimeException(INVALID_CITY);
		
		return new Address(street, city, zipCode);
	}
	
	private Address (String street, String city, String zipCode) {
		this.street = street;
		this.city = city;
		this.zipCode = zipCode;
	}

	public String getCity() {
		return city;
	}

	public String getZipCode() {
		return zipCode;
	}
}
